### PR TITLE
[exporter/elasticsearch] fix not exporting span events

### LINKED ISSUE #18479: [exporter/elasticsearch] Elasticsearch Exporter Not Exporting Span Events
### Component(s)

exporter/elasticsearch

### What happened?

The exported span record doesn't contain: **span events** informaton:

C# emitting code:

```
        [HttpGet]
        [Route("Random")]
        public IEnumerable<int> Get([FromQuery] int count)
        {
            using var span = _tracer.StartActiveSpan("GetRandoms");

            span.AddEvent("**Generating Random Numbers**", ->missing
                new SpanAttributes(new Dictionary<string, object> { { "Count", count } }));

            var randoms = Enumerable.Range(1, count).Select(_ => Random.Next(1, 100)).ToArray();

            span.AddEvent("Random Numbers Were Generated", ->missing
                new SpanAttributes(new Dictionary<string, object>
                    { { "Count", count }, { "Numbers", string.Join(",", randoms) } }));

            span.SetStatus(Status.Ok);

            return randoms;
        }
```

```
{
    "_index": "otel_trace_index",
    "_id": "AdJ-NIYBXAUPn3c51_CD",
    "_version": 1,
    "_seq_no": 6,
    "_primary_term": 1,
    "found": true,
    "_source": {
        "@timestamp": "2023-02-09T04:45:53.221066600Z",
        "EndTimestamp": "2023-02-09T04:45:53.221324100Z",
        "Kind": "SPAN_KIND_INTERNAL",
        "Link": "[]",
        "Name": "GetRandoms",
        "ParentSpanId": "6bd3ad5ac67e7db8",
        "Resource.service.instance.id": "064dd191-d7ee-4b12-abe8-60593845c149",
        "Resource.service.name": "RootService",
        "SpanId": "b048c67526d65cea",
        "TraceId": "5dfff6df3a0cdaf0e34637c0ab64f918",
        "TraceStatus": 1
    }
}
```

### Collector version

v0.70.0

### Environment information

_No response_

### OpenTelemetry Collector configuration

```yaml
receivers:
  otlp:
    protocols:
      grpc:
        endpoint: '0.0.0.0:4317'
      http:
        endpoint: '0.0.0.0:4318'
        cors:
          allowed_origins:
            - 'http://*'
            - 'https://*'

processors:
  batch:
    send_batch_size: 1000
    timeout: 1s
  memory_limiter:
    limit_mib: 1500
    spike_limit_mib: 512
    check_interval: 5s

extensions:
  health_check:
  pprof:
  zpages:

exporters:
  elasticsearch/trace:
    endpoints: [http://es-container:9200]
    traces_index: otel_trace_index
  elasticsearch/log:
    endpoints: [http://es-container:9200]
    logs_index: otel_log_index
      
service:
  extensions: [pprof, zpages, health_check]
  pipelines:
    traces:
      receivers:
        - otlp
      processors:
        - memory_limiter
        - batch
      exporters:
        - elasticsearch/trace
    logs:
      receivers:
        - otlp
      processors:
        - memory_limiter
        - batch
      exporters:
        - elasticsearch/log
```

### Log output

```shell
N/A
```

### Additional context

N/A

### PR DESCRIPTION
**Pull request by JaredTan95**
fix https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18479
Fix elasticsearch exporter not exporting span events

**Description:** <Describe what has changed.>

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: exporter/elasticsearchexporter/internal/objmodel/objmodel.go [hunk 1]
```diff
@@ -175,6 +176,14 @@ func (doc *Document) AddAttribute(key string, attribute pcommon.Value) {
 	}
 }
 
+// AddEvents converts and adds span events to the document.
+func (doc *Document) AddEvents(key string, events ptrace.SpanEventSlice) {
+	for i := 0; i < events.Len(); i++ {
+		e := events.At(i)
+		doc.AddAttributes(flattenKey(key, e.Name()), e.Attributes())
```
Comment (dmitryax) [inline exporter/elasticsearchexporter/internal/objmodel/objmodel.go:183]: What about the event timestamp? That's pretty important information

Comment (JaredTan95) [reply exporter/elasticsearchexporter/internal/objmodel/objmodel.go:183]: aha, Indeed, I will add `time_unix_nano` into attributes.

File: exporter/elasticsearchexporter/internal/objmodel/objmodel.go [hunk 2]
```diff
@@ -175,6 +176,15 @@ func (doc *Document) AddAttribute(key string, attribute pcommon.Value) {
 	}
 }
 
+// AddEvents converts and adds span events to the document.
+func (doc *Document) AddEvents(key string, events ptrace.SpanEventSlice) {
+	for i := 0; i < events.Len(); i++ {
+		e := events.At(i)
+		doc.Add(flattenKey(key, e.Name()+".time"), TimestampValue(e.Timestamp().AsTime()))
+		doc.AddAttributes(flattenKey(key, e.Name()), e.Attributes())
```
Comment (TylerHelmuth) [inline exporter/elasticsearchexporter/internal/objmodel/objmodel.go:184]: Do you care about `dropped_attributes_count`?

File: exporter/elasticsearchexporter/internal/objmodel/objmodel.go [hunk 3]
```diff
@@ -175,6 +176,15 @@ func (doc *Document) AddAttribute(key string, attribute pcommon.Value) {
 	}
 }
 
+// AddEvents converts and adds span events to the document.
+func (doc *Document) AddEvents(key string, events ptrace.SpanEventSlice) {
+	for i := 0; i < events.Len(); i++ {
+		e := events.At(i)
+		doc.Add(flattenKey(key, e.Name()+".time"), TimestampValue(e.Timestamp().AsTime()))
```
Comment (dmitryax) [inline exporter/elasticsearchexporter/internal/objmodel/objmodel.go:183]: Why time and not timestamp?

Comment (JaredTan95) [reply exporter/elasticsearchexporter/internal/objmodel/objmodel.go:183]: currently, any time in span storage with `date time` format in Elasticsearch json documents.  and I think `data time` format is more readable.

File: exporter/elasticsearchexporter/internal/objmodel/objmodel.go (hunk 2 above)
Comment (JaredTan95) [reply exporter/elasticsearchexporter/internal/objmodel/objmodel.go:184]: I don't care `dropped_attributes_count ` actually, But, I will add it here to make the event store more complete.

**Comment by TylerHelmuth** [comment]
Please get the latest from main to fix the CI issues.

**Comment by JaredTan95** [comment]
> Please get the latest from main to fix the CI issues.

yes, I'm also waiting the main branch get this CI fixed

**Review by TylerHelmuth — APPROVED** [review]

File: exporter/elasticsearchexporter/internal/objmodel/objmodel.go (hunk 3 above)
Comment (dmitryax) [reply exporter/elasticsearchexporter/internal/objmodel/objmodel.go]: data time representation is good. My Q is about naming conventions. Do we have any policy on translating the OTel schema to the elastic? So far, I see that proto representation in CamelCase is used. Here it's a bit different, especially for `dropped_attributes_count`.

Do you think it's something that needs to be defined and updated in the exporter? If so, it'd be good to have an issue for that.

Comment (dmitryax) [reply exporter/elasticsearchexporter/internal/objmodel/objmodel.go]: Looks like we have some rules in the spac https://github.com/open-telemetry/opentelemetry-specification/blob/773ee656f92c7f591f2fd9c38df82c264a15184d/specification/logs/data-model-appendix.md#elastic-common-schema which still need to be completed, but at least we have something to start with

Comment (JaredTan95) [reply exporter/elasticsearchexporter/internal/objmodel/objmodel.go]: thx, I got you, I reviewed https://github.com/open-telemetry/opentelemetry-specification/blob/773ee656f92c7f591f2fd9c38df82c264a15184d/specification/logs/data-model-appendix.md#elastic-common-schema and found that  `dropped_attributes_count` as nested fields is consistent with the style with `cloud.availability_zone`, `error.stack_trace`  in the log model. 

BTW, I agree with we need to have an issue as a good start. But, this issue should be placed in `opentelemetry-specification` repo or `otel-col-contrib`? I guess `opentelemetry-specification`? And It should be completed in another PR?

**Comment by atoulme** [comment]
Please rebase.

**Comment by JaredTan95** [comment]
> Please rebase.

thx, done.

File: exporter/elasticsearchexporter/internal/objmodel/objmodel.go (hunk 3 above)
Comment (dmitryax) [reply exporter/elasticsearchexporter/internal/objmodel/objmodel.go]: First, we need an issue in contrib to align the schema used in elastic search with what we already have in the spec (or at least as close as possible). Anything that is unclear or missing in the spec should be updated. Fill free to raise issues or PRs there too

File: exporter/elasticsearchexporter/internal/objmodel/objmodel.go (hunk 2 above)
Comment (dmitryax) [reply exporter/elasticsearchexporter/internal/objmodel/objmodel.go:174]: I would recommend dropping  `dropped_attributes_count` as it's pretty misaligned with the recommendation for the OTLP-Elastic schema translations in the spec. It'll be less to break going forward

File: exporter/elasticsearchexporter/internal/objmodel/objmodel.go (hunk 3 above)
Comment (JaredTan95) [reply exporter/elasticsearchexporter/internal/objmodel/objmodel.go]: Yes, I track it with the issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/22137, and I will PR ASAP.

Comment (dmitryax) [reply exporter/elasticsearchexporter/internal/objmodel/objmodel.go]: Thanks!

**Review by dmitryax — APPROVED** [review]

**Pull request merged**